README.md
<img src="https://gist.githubusercontent.com/joaolucasl/00f53024cecf16410d5c3212aae92c17/raw/1789a2131ee389aeb44e3a9d5333f59cfeebc089/moip-icon.png" align="right" />

# Moip Java SDK
> Moip API v2 Java SDK for client integration.

[![Build Status](https://travis-ci.org/moip/moip-sdk-java.svg?branch=master)](https://travis-ci.org/moip/moip-sdk-java)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/59c15b9d4e35440c8e1d2810c0509836)](https://www.codacy.com/app/rodrigo-saito/moip-sdk-java)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg)](https://github.com/moip/moip-sdk-java/blob/master/LICENSE)
[![Slack](https://slackin-cqtchmfquq.now.sh/badge.svg)](https://slackin-cqtchmfquq.now.sh/)

<details>
    <summary>Index</summary>

* [Simple flow](#simple-flow)
  * [Setup](#setup)
    * [Setting the authentication](#setting-the-authentication)
    * [Client](#client)
    * [Instantiate API](#instantiate-api)
  * [Create customer](#create-customer)
  * [Create order](#create-order)
  * [Create Payment](#create-payment)
  * [Other examples](#other-examples)
* [Exceptions treatment](#exceptions-treatment)
* [Moip documentation](#moip-documentation)
* [Getting help](#getting-help)
* [Contributing](#contributing)

</details>

## Require

Java `v1.7ˆ` ![java-cup](https://user-images.githubusercontent.com/32847427/37723265-b8441610-2d0c-11e8-8238-ab27df829a13.png)

## Installing

### Gradle

![Maven-central](https://img.shields.io/maven-central/v/br.com.moip/java-sdk.svg)

Add the following dependency to `build.gradle` in the project:

```gradle
compile group: 'br.com.moip', name: 'java-sdk', version: 'x.y.z'
```

### Maven

![Maven-central](https://img.shields.io/maven-central/v/br.com.moip/java-sdk.svg)

Add the following dependency to `pom.xml` in the project:

```xml
<dependency>
    <groupId>br.com.moip</groupId>
    <artifactId>java-sdk</artifactId>
    <version>x.y.z</version>
</dependency>

```

## Simple flow

This step by step will exemplify the integration flow with simple usage examples.

### 1. Setup
Before making requests to Moip API it's necessary make a **setup**, defining the environment and the authentication that will be used.

#### 1.1 Setting the authentication
There are two ways to authenticate the request, some endpoints require a "highest authorization level", it will depend on the endpoint and type of request.

#### By BasicAuth
The following set will generate a hash `Base64` with your Moip account **token** and **key** to authenticate.
```java
Authentication auth = new BasicAuth("TOKEN", "SECRET");
```

> :bulb: If you don't know how to get your **token** and **key**, click [here](https://conta-sandbox.moip.com.br/configurations/api_credentials) (you must be logged in).

#### By OAuth
The following set will create an OAuth authentication object.

> :bulb: Click [here](https://dev.moip.com.br/v2.0/reference#autenticacao-mp) to know how to get your token OAuth.

```java
Authentication auth = new OAuth("TOKEN_OAUTH");
```

#### 1.2 Client
After defining your authentication method, you have to set the **client** specifying the environment where you want to run your application. To set the client, choose the environment and pass the authentication (previously settled) as argument.

##### Sandbox
The test environment. You can use this to simulate all of your business scenarios.

```java
Client client = new Client(Client.SANDBOX, auth);
```

##### Production
**_"The environment of truth"_** :eyes:. This is the environment where the real transactions run.

```java
Client client = new Client(Client.PRODUCTION, auth);
```

> :bulb: Before going to production, you need to request homologation of your application [here](https://dev.moip.com.br/page/homologacao-api-v2).

#### 1.3 Instantiate API
To complete the setup, you have to create an API instance, passing the client (previously settled) as argument.

```java
API api = new API(client);
```

> :white_check_mark: Check the [setup functional example](https://github.com/moip/moip-sdk-java/tree/master/src/test/java/br/com/moip/examples/setup/Setup.java).

### 2. Create customer
With the setup created, you can make requests to Moip API. To start the basic e-commerce flow you need to create a customer. After all, it's whom will order your products or services.

```java
Customer customer = api.customer().create(new CustomerRequest()
    .ownId("CUS-" + System.currentTimeMillis())
    .fullname("Jose da Silva")
    .email("josedasilva@email.com")
    .birthdate(new ApiDateRequest().date(new Date()))
    .taxDocument(TaxDocumentRequest.cpf("22222222222"))
    .phone(new PhoneRequest().setAreaCode("11").setNumber("55443322"))
    .shippingAddressRequest(new ShippingAddressRequest()
        .street("Avenida Faria Lima")
        .streetNumber("3064")
        .complement("12 andar")
        .city("São Paulo")
        .state("SP")
        .district("Itaim")
        .country("BRA")
        .zipCode("01452-000")
    )
);
```
> [:white_check_mark:](https://github.com/moip/moip-sdk-java/tree/master/src/test/java/br/com/moip/examples/CustomerExamples.java) _Click on the icon for more customer functional examples._
> Read more about customer on [API reference](https://dev.moip.com.br/v2.0/reference#clientes-ec).

###  3. Create order
Customer created! It's buy time! :tada:

```java
Order createdOrder = api.order().create(new OrderRequest()
    .ownId("ORD-" + System.currentTimeMillis())
    .amount(new OrderAmountRequest()
        .currency("BRL")
        .subtotals(new SubtotalsRequest()
            .shipping(1000)
            .addition(100)
            .discount(500)
        )
    )
    .addItem("Nome do produto 1", 1, "Mais info...", 100)
    .addItem("Nome do produto 2", 2, "Mais info...", 200)
    .addItem("Nome do produto 3", 3, "Mais info...", 300)
    .customer(new CustomerRequest()
        .id("CUSTOMER_ID")
    )
    .addReceiver(new ReceiverRequest()
        .secondary("MOIP_ACCOUNT_ID", new AmountRequest().percentual(50), false)
    )
);
```
> [:white_check_mark:](https://github.com/moip/moip-sdk-java/tree/master/src/test/java/br/com/moip/examples/OrderExamples.java) _Click on the icon for more order functional examples._
> Read more about order on [API reference](https://dev.moip.com.br/v2.0/reference#pedidos-ec).

### 4. Create Payment
Alright! Do you have all you need? So, lets pay this order. :moneybag:

```java
Payment createdPayment = api.payment().create(new PaymentRequest()
    .orderId("ORDER_ID")
    .installmentCount(1)
    .fundingInstrument(new FundingInstrumentRequest()
        .creditCard(new CreditCardRequest()
            .number("5555666677778884")
            .cvc(123)
            .expirationMonth("12")
            .expirationYear("20")
            .holder(new HolderRequest()
                .fullname("Jose Portador da Silva")
                .birthdate("1988-10-10")
                .phone(new PhoneRequest()
                    .setAreaCode("11")
                    .setNumber("55667788")
                )
                .taxDocument(TaxDocumentRequest.cpf("22222222222"))
            )
            .store(true)
        )
    )
);
```
> [:white_check_mark:](https://github.com/moip/moip-sdk-java/tree/master/src/test/java/br/com/moip/examples/PaymentExamples.java) _Click on the icon for more payment functional examples._
> Read more about payment on [API reference](https://dev.moip.com.br/v2.0/reference#pagamentos-ec).

### Other examples
If you want to see other functional examples, check this folder. [:file_folder:](https://github.com/moip/moip-sdk-java/tree/master/src/test/java/br/com/moip/examples/)

## Exceptions treatment
| errors | cause | status |
| :---: | :---: | :---: |
| UnautorizedException | to authentication errors | == 401 |
| ValidationException | to validation errors | >= 400 && <= 499 (except 401) |
| UnexpectedException | to unexpected errors | >= 500 |

> :warning: To catch these errors, use the bellow treatment:

```java
try {
 Payment createdPayment = api.payment().create(
        //...
    );
} catch(UnauthorizedException e) {
  // StatusCode == 401
} catch(UnexpectedException e) {
  // StatusCode >= 500
} catch(ValidationException e) {
  // StatusCode entre 400 e 499 (exceto 401)
}
```

## Moip documentation

### Docs
To stay up to date about the **Moip Products**, check the [documentation](https://dev.moip.com.br/v2.0/docs).

### References
Read more about the **Moip APIs** in [API reference](https://dev.moip.com.br/v2.0/reference).

## Getting help
We offer many ways to contact us, so if you have a question, do not hesitate, talk to us whatever you need. For questions about API or business rules, contact us by [support](https://dev.moip.com.br/v2.0/) or [slack](https://slackin-cqtchmfquq.now.sh/):slack:. But, if you have a question or suggestion about the SDK, feel free to open an **issue** or **pull request**.

## Contributing
Do you have an enhancement suggest or found something to fix? Go ahead, help us and let your mark on Moip, open **pull requests** and **issues** against this project. If you want to do it, please read the `CONTRIBUTING.md` to be sure everyone follows the same structure and planning of the project. Remember, we :heart: contributions. :rocket:


CHANGELOG.md
<a name="v4.3.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v4.2.0...v4.3.0) (2019-12-11)


### Features
* **CreditCard**: adds `getExpirationMonth()` and `getExpirationYear()` methods ([ed1b0a3](https://github.com/moip/moip-sdk-java/commits/ed1b0a3226f3f3766b81a074b8ae75e31b7c184e))


<a name="v4.2.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v4.1.1...v4.2.0) (2018-05-02)


### Bug Fixes

* **Client**: apply fixes to JSON translation ([bf40270]())

### Refactor

* **Account**: fix account treatment ([4d50db3]())

### Docs

* **Readme**
  * Adding slack link ([2ff4443]())
  * complete and translate the documentation ([d3b63b5]())

* **Examples**
  * adds request examples for all SDK features ([95e6a0a]())

* **Contributing**: adds CONTRIBUTING.md ([c2b2de5]())


<a name="v4.1.1"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v4.1.0...v4.1.1) (2018-03-02)


### Bug Fixes

* **Client**: fix a bug in method `doRequest()` ([]())


<a name="v4.1.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v4.0.0...v4.1.0) (2018-02-27)


### Features

* **Payment**: adds `getEscrowId()` and `getEscrow()` methods ([e51d6cc](https://github.com/moip/moip-sdk-java/commits/e51d6cc))
* **Entry**: adds entry feature and your own resources ([8cdb098](https://github.com/moip/moip-sdk-java/commits/8cdb098))
* **Balances**: adds feature to get Moip Account's Balances ([ff6f41b](https://github.com/moip/moip-sdk-java/commits/ff6f41b))
* **Refund**: adds list order refunds and payment refunds ([b3a0844](https://github.com/moip/moip-sdk-java/commits/b3a0844))

### Refactor

* **Client**:
  * adds a new get with new header ([41edb85](https://github.com/moip/moip-sdk-java/commits/41edb85))
  * adjust Accept values ([3fe45c7](https://github.com/moip/moip-sdk-java/commits/3fe45c7))
* **API**: fix imports ([817eae6](https://github.com/moip/moip-sdk-java/commits/817eae6))
* **Payment**: adds statementDescriptor ([15ff550](https://github.com/moip/moip-sdk-java/commits/15ff550))
* **RefundAPI**: change condition of `list()` method ([55ff9d0](https://github.com/moip/moip-sdk-java/commits/55ff9d0))
* **Order**: add missing response elements ([b8ba4bc](https://github.com/moip/moip-sdk-java/commits/b8ba4bc))

### Tests

* **Balances**: adds tests to balances ([4c8e251](https://github.com/moip/moip-sdk-java/commits/4c8e251))
* **Payment**/**Multipayment**: add tests to statement descriptor ([03d8f4f](https://github.com/moip/moip-sdk-java/commits/03d8f4f))
* **Refund**: adds tests to list refunds ([080fd10](https://github.com/moip/moip-sdk-java/commits/080fd10))
* **Order**: adds tests to cover all response attributes ([c47071f](https://github.com/moip/moip-sdk-java/commits/c47071f))


<a name="v4.0.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v3.4.0...v3.5.0) (2017-12-05)


### Features

* **Account:** added return accessToken and setPassword ([a078943](https://github.com/moip/moip-sdk-java/commits/a078943))
* **Account:** created AccountLinks class and added some account creation tests ([1e5632b](https://github.com/moip/moip-sdk-java/commits/1e5632b))
* **Order:** return checkout preferences and checkout links ([26e96de](https://github.com/moip/moip-sdk-java/commits/26e96de))
* **Order:** create alternatives methods to get the checkout links and add some tests ([d9b17ce](https://github.com/moip/moip-sdk-java/commits/a078943))
* **Transfer:** adds create and revert a transfer ([1f9357a](https://github.com/moip/moip-sdk-java/commits/1f9357a))
* **Transfer:** adds MoipAccountRequest, fix some parameters and fix Transfer examples ([5f274c5](https://github.com/moip/moip-sdk-java/commits/5f274c5))
* **Multipayment** adds Checkout instance to MultipaymentLinks ([dad95a2](https://github.com/moip/moip-sdk-java/commits/dad95a2))

### Refactor

* **Order:** fixed parameter `addiction` to `addition` ([42a08c3](https://github.com/moip/moip-sdk-java/commits/42a08c3))


<a name="v3.4.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v3.3.0...v3.4.0) (2017-11-08)


### Features

* **Order:** adding OrderStatus to Order object ([71a42d8](https://github.com/moip/moip-sdk-java/commit/71a42d8))
* **CreditCardRequest:** Add id property, getter and setter ([bb59c12](https://github.com/moip/moip-sdk-java/commit/bb59c12))
* **Amount:** Added fees, refunds, liquid, currency properties and getters for them ([e90fce0](https://github.com/moip/moip-sdk-java/commit/e90fce0))
* **BankAccount** Add method to delete a bank account. ([113ff51](https://github.com/moip/moip-sdk-java/commit/113ff51))
* **BankAccount** Add method to update a bank account. ([b99fd12](https://github.com/moip/moip-sdk-java/commit/b99fd12))


<a name="v3.3.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v3.2.0...v3.3.0) (2017-10-10)


### Features

* **Multipayment:** Added method 'cancelPreAuthorized' to cancel pre authorized payments ([be2fc6c](https://github.com/moip/moip-sdk-java/commit/be2fc6c))
* **Multipayment:** Added method 'capture' ([0a634c3](https://github.com/moip/moip-sdk-java/commit/0a634c3))
* **Order:** added method to get an order list ([8602c59](https://github.com/moip/moip-sdk-java/commit/8602c59))



<a name="v3.2.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v3.1.0...v3.2.0) (2017-09-28)

### Features

* **BankAccount:** Added accountCheckNumber property to BankAccount([8ab448b](https://github.com/moip/moip-sdk-java/commit/8ab448b))


<a name="v3.1.0"></a>
# [](https://github.com/moip/moip-sdk-java/compare/v3.0.0...v3.1.0) (2017-09-18)


### Bug Fixes

* **ConnectAPI:** Fix a bug in method to generate authorize URL ([851aa2a](https://github.com/moip/moip-sdk-java/commit/851aa2a))

### Features

* Create structure to HATEOAS links in Payment and fix toString methods to some Link Resources ([cf6c46f](https://github.com/moip/moip-sdk-java/commit/cf6c46f))


<a name="v3.0.0"></a>
# [3.0.0](https://github.com/moip/moip-sdk-java/compare/v2.0.0-RC6...v3.0.0) (2017-09-11)



src/main/java/br/com/moip/Client.java
package br.com.moip;

import br.com.moip.authentication.Authentication;
import br.com.moip.exception.MoipException;
import br.com.moip.exception.UnauthorizedException;
import br.com.moip.exception.UnexpectedException;
import br.com.moip.exception.ValidationException;
import br.com.moip.resource.Errors;
import br.com.moip.ssl.SSLSupport;
import br.com.moip.util.GsonFactory;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import javax.net.ssl.HttpsURLConnection;
import java.io.DataOutputStream;
import java.io.OutputStreamWriter;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import org.apache.http.entity.ContentType;

import static br.com.moip.util.DataHelper.jsonToUrlEncodedString;

public class Client {

    private static final Logger LOGGER = LoggerFactory.getLogger(Client.class);
    public static final String PRODUCTION = "https://api.moip.com.br";
    public static final String SANDBOX = "https://sandbox.moip.com.br";
    public static final String CONNECT_PRODUCTION = "https://connect.moip.com.br";
    public static final String CONNECT_SANDBOX = "https://connect-sandbox.moip.com.br";
    private static String USER_AGENT;
    private final String RESPONSE_BODY_400 = "400";
    private final String RESPONSE_BODY_404 = "404";

    static {
        try {
            InputStream inputStream = Client.class.getResourceAsStream("/moipJavaSDK.properties");
            Properties properties = new Properties();
            properties.load(inputStream);

            USER_AGENT = properties.getProperty("userAgent");
        } catch (Exception e) {
            USER_AGENT = "MoipJavaSDK/UnknownVersion (+https://github.com/moip/moip-sdk-java/)";
        }
    }

    private final String endpoint;
    private final Authentication authentication;
    private final Gson gson;

    public Client(final String endpoint, final Authentication authentication) {
        this.endpoint = endpoint;
        this.authentication = authentication;
        this.gson = GsonFactory.gson();
    }

    public <T> T post(final String path, final Class<T> type) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("POST").path(path).type(type).contentType(ContentType.APPLICATION_JSON);
        return doRequest(props);
    }

    public <T> T post(final String path, final Object object, final Class<T> type) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("POST").path(path).object(object).type(type).contentType(ContentType.APPLICATION_JSON);
        return doRequest(props);
    }

    public <T> T post(final String path, final Object object, final Class<T> type, ContentType contentType) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("POST").path(path).object(object).type(type).contentType(contentType);
        return doRequest(props);
    }

    public <T> T put(final String path, final Object object, final Class<T> type) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("PUT").path(path).object(object).type(type).contentType(ContentType.APPLICATION_JSON);
        return doRequest(props);
    }

    public <T> T get(String path, Class<T> type) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("GET").path(path).type(type).contentType(ContentType.APPLICATION_JSON);
        return doRequest(props);
    }

    public <T> T get(String path, Class<T> type, String acceptVersion) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("GET").path(path).type(type).contentType(ContentType.APPLICATION_JSON).accept(acceptVersion);
        return doRequest(props);
    }

    public <T> T delete(String path, Class<T> type) {
        RequestProps props = RequestPropsBuilder.requestPropsBuilder().method("DELETE").path(path).object(null).type(type).contentType(ContentType.APPLICATION_JSON);
        return doRequest(props);
    }

    private <T> T doRequest(final RequestProps requestProps) {
        try {
            URL url = new URL(endpoint + requestProps.path);

            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestProperty("User-Agent", USER_AGENT);
            conn.setRequestProperty("Content-type", requestProps.contentType.getMimeType());
            if (requestProps.accept != null) conn.setRequestProperty("Accept", requestProps.accept);

            conn.setRequestMethod(requestProps.method);

            // Disable TLS 1.0
            if (conn instanceof HttpsURLConnection) {
                ((HttpsURLConnection) conn).setSSLSocketFactory(new SSLSupport());
            }

            if (authentication != null) {
                authentication.authenticate(conn);
            }

            LOGGER.debug("---> {} {}", requestProps.method, conn.getURL().toString());
            logHeaders(conn.getRequestProperties().entrySet());

            if (requestProps.object != null) {
                conn.setDoOutput(true);
                String body = getBody(requestProps.object, requestProps.contentType);

                LOGGER.debug("{}", body);

                DataOutputStream wr = new DataOutputStream(conn.getOutputStream());
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(wr, "UTF-8"));
                writer.write(body);
                writer.close();
                wr.flush();
                wr.close();
            }

            LOGGER.debug("---> END HTTP");

            int responseCode = conn.getResponseCode();

            LOGGER.debug("<--- {} {}", responseCode, conn.getResponseMessage());
            logHeaders(conn.getHeaderFields().entrySet());

            StringBuilder responseBody = new StringBuilder();

            responseBody = responseBodyTreatment(responseBody, responseCode, conn);

            LOGGER.debug("{}", responseBody.toString());
            LOGGER.debug("<-- END HTTP ({}-byte body)", conn.getContentLength());

            return gson.fromJson(responseBody.toString(), requestProps.<T>getType());
        } catch (IOException | KeyManagementException | NoSuchAlgorithmException e) {
            throw new MoipException("Error occurred connecting to Moip API: " + e.getMessage(), e);
        }
    }

    private StringBuilder responseBodyTreatment(StringBuilder responseBody, int responseCode, HttpURLConnection conn) {

        try {

            if (responseCode >= 200 && responseCode < 299) {
                responseBody = readBody(conn.getInputStream());
            }

            if (responseCode == 401) {
                throw new UnauthorizedException();
            }

            if (responseCode >= 400 && responseCode < 499) {
                responseBody = readBody(conn.getErrorStream());
                LOGGER.debug("API ERROR {}", responseBody.toString());

                Errors errors = new Errors();

                try {

                    errors = gson.fromJson(responseBody.toString(), Errors.class);

                } catch (Exception e) {

                    LOGGER.debug("There was not possible cast the JSON to object");
                }

                throw new ValidationException(responseCode, conn.getResponseMessage(), errors);
            }

            if (responseCode >= 500) {
                throw new UnexpectedException();
            }

        } catch (IOException e) {
            throw new MoipException("Error occurred connecting to Moip API: " + e.getMessage(), e);
        }

        return responseBody;
    }

    private void logHeaders(Set<Map.Entry<String, List<String>>> entries) {
        for (Map.Entry<String, List<String>> header : entries) {
            if (header.getKey() != null) {
                LOGGER.debug("{}: {}", header.getKey(), header.getValue());
            }
        }
    }

    private StringBuilder readBody(final InputStream inputStream) throws IOException {
        StringBuilder body = new StringBuilder();
        BufferedReader in = new BufferedReader(new InputStreamReader(inputStream));

        String inputLine;
        while ((inputLine = in.readLine()) != null) {
            body.append(inputLine);
        }
        in.close();

        return body;
    }

    private String getBody(Object object, ContentType contentType) {
        if (contentType == ContentType.APPLICATION_FORM_URLENCODED) {
            return jsonToUrlEncodedString((JsonObject) new JsonParser().parse(gson.toJson(object)));
        }

        return gson.toJson(object);
    }

    public Authentication getAuthentication() {
        return authentication;
    }

    public String getEndpoint() {
        return endpoint;
    }

    private static class RequestProps {

        protected String method;
        protected String path;
        protected Object object;
        protected Class type;
        protected ContentType contentType;
        protected String accept;

        public RequestProps() {}

        public String getMethod() { return method; }

        public String getPath() { return path; }

        public Object getObject() { return object; }

        public <T> Class<T> getType() { return type; }

        public ContentType getContentType() { return contentType; }

        public String getAccept() { return accept; }
    }

    private static class RequestPropsBuilder extends RequestProps {

        public static RequestPropsBuilder requestPropsBuilder() {
            return new RequestPropsBuilder();
        }

        public RequestPropsBuilder method(String method) {
            this.method = method;
            return this;
        }

        public RequestPropsBuilder path(String path) {
            this.path = path;
            return this;
        }

        public RequestPropsBuilder object(Object object) {
            this.object = object;
            return this;
        }

        public RequestPropsBuilder type(Class type) {
            this.type = type;
            return this;
        }

        public RequestPropsBuilder contentType(ContentType contentType) {
            this.contentType = contentType;
            return this;
        }

        public RequestPropsBuilder accept(String acceptVersion) {
            this.accept = acceptBuilder(acceptVersion);
            return this;
        }

        public String acceptBuilder(String version) {
            String value = "application/json";
            if(version == "2.1") value += ";version=" + version;
            return value;
        }
    }

}